### PR TITLE
Make standalone bin load path relative to Bundler.root rather than CWD.

### PR DESCRIPTION
**Pull request by glennpratt**
See #4144

### CODE REVIEW

**Comment by glennpratt** [comment]
Issue #4144

**Comment by segiddins** [comment]
Please add a spec for whatever it is you're fixing?

File: lib/bundler/installer.rb [hunk 1]
```diff
@@ -130,12 +130,13 @@ def generate_bundler_executable_stubs(spec, options = {})
     def generate_standalone_bundler_executable_stubs(spec)
       # double-assignment to avoid warnings about variables that will be used by ERB
       bin_path = Bundler.bin_path
+      standalone_full_path = Pathname(File.join(Bundler.root, Bundler.settings[:path]))
```
Comment (segiddins) [inline lib/bundler/installer.rb]: using `File.join` isn't necessary, since `Bundler.root` is already a `Pathname`

**Comment by glennpratt** [comment]
@segiddins I'll try to come up with something. The bug appears to be exposed while using multiple jobs (threads) and isn't consistent. I could probably force the issue with mocks, but I'm not sure how good of a test that is.

File: lib/bundler/installer.rb (hunk 1 above)
Comment (glennpratt) [inline lib/bundler/installer.rb]: Thanks, pushed a change for that.

File: .rubocop.yml [hunk 2]
```diff
@@ -4,6 +4,7 @@ inherit_from:
 AllCops:
   TargetRubyVersion: 1.9
   Exclude:
+    - bundle/**/*
```
Comment (segiddins) [inline .rubocop.yml]: this change should be unecessary

Comment (glennpratt) [inline .rubocop.yml]: Sorry, I didn't intend to add it, though without it, trying to develop just using the rake tasks:

```
rake rubocop
<snip>
241 files inspected, 9403 offenses detected

7 errors occurred:
An error occurred while Performance/RedundantBlockCall cop was inspecting /Users/glenn.pratt/Code/bundler/bundle/ruby/2.2.0/gems/rack-1.6.4/lib/rack/nulllogger.rb.
An error occurred while Performance/RedundantBlockCall cop was inspecting /Users/glenn.pratt/Code/bundler/bundle/ruby/2.2.0/gems/rack-1.6.4/lib/rack/nulllogger.rb.
An error occurred while Performance/RedundantBlockCall cop was inspecting /Users/glenn.pratt/Code/bundler/bundle/ruby/2.2.0/gems/rack-1.6.4/lib/rack/nulllogger.rb.
An error occurred while Performance/RedundantBlockCall cop was inspecting /Users/glenn.pratt/Code/bundler/bundle/ruby/2.2.0/gems/rack-1.6.4/lib/rack/nulllogger.rb.
An error occurred while Performance/RedundantBlockCall cop was inspecting /Users/glenn.pratt/Code/bundler/bundle/ruby/2.2.0/gems/rack-1.6.4/lib/rack/nulllogger.rb.
An error occurred while Performance/RedundantBlockCall cop was inspecting /Users/glenn.pratt/Code/bundler/bundle/ruby/2.2.0/gems/rack-1.6.4/lib/rack/nulllogger.rb.
An error occurred while Performance/RedundantBlockCall cop was inspecting /Users/glenn.pratt/Code/bundler/bundle/ruby/2.2.0/gems/rack-1.6.4/lib/rack/nulllogger.rb.
Errors are usually caused by RuboCop bugs.
Please, report your problems to RuboCop's issue tracker.
Mention the following information in the issue report:
0.36.0 (using Parser 2.3.0.1, running on ruby 2.2.3 x86_64-darwin14)
RuboCop failed!
```

**Comment by segiddins** [comment]
Make sure to use `bin/rake`

**Comment by glennpratt** [comment]
@segiddins same result with `bin/rake rubocop`. Up to date with master.

**Comment by segiddins** [comment]
In that case, just `git clean -xdf ` before you run the specs, I guess?

**Comment by segiddins** [comment]
We can't merge this until test coverage is added and the rubocop config change is backed out, sorry.

**Comment by glennpratt** [comment]
No problem, I'm doing that now.

**Comment by glennpratt** [comment]
@segiddins I created a simple test though it doesn't fail with fake local gems anyway. It does fail using Rubygems and multiple jobs about 1/8th of the time.

**Comment by segiddins** [comment]
@glennpratt you can `chdir` in your tests, then joining with root and `expand_path` would have different values

**Comment by glennpratt** [comment]
@segiddins well, I tried that but it exposed an apparently different bug (`bundler/setup.rb` ends up in the current directory rather than `root/bundle/bundler/setup.rb`).

**Comment by glennpratt** [comment]
On master, running from the subdirectory bob results in:

```
tree tmp/bundled_app/
tmp/bundled_app/
├── Gemfile
├── Gemfile.lock
├── bin
│   ├── rails
│   └── rake
├── bob
│   └── bundle
│       └── bundler
│           └── setup.rb
└── bundle
    └── ruby
        └── 2.2.0
            ├── bin
            │   ├── rails
            │   └── rake
...
```

The binstubs also don't work:
```
/Users/glenn.pratt/.rvm/rubies/ruby-2.2.3/lib/ruby/2.2.0/rubygems/core_ext/kernel_require.rb:54:in `require': cannot load such file -- rails (LoadError)
	from /Users/glenn.pratt/.rvm/rubies/ruby-2.2.3/lib/ruby/2.2.0/rubygems/core_ext/kernel_require.rb:54:in `require'
	from /Users/glenn.pratt/Code/bundler/tmp/bundled_app/bundle/ruby/2.2.0/gems/rails-2.3.2/bin/rails:2:in `<top (required)>'
	from /Users/glenn.pratt/Code/bundler/tmp/bundled_app/bin/rails:12:in `load'
	from /Users/glenn.pratt/Code/bundler/tmp/bundled_app/bin/rails:12:in `<main>'
```

So, I'm not exactly sure the correct behavior, but I can put a pending test in and a todo.

**Comment by glennpratt** [comment]
@segiddins updated.

File: spec/install/gems/standalone_spec.rb [hunk 3]
```diff
@@ -272,5 +272,36 @@
...
       end
     end
+
+    it "creates stubs with the correct load path" do
+      extension_line = File.read(bundled_app("bin/rails")).each_line.find {|line| line.include? "$:.unshift" }.strip
+      expect(extension_line).to eq "$:.unshift File.expand_path '../../bundle', __FILE__"
+    end
+  end
+
+  describe 'with --binstubs run in a subdirectory' do
```
Comment (segiddins) [inline spec/install/gems/standalone_spec.rb]: double quotes

Comment (glennpratt) [inline spec/install/gems/standalone_spec.rb]: fixed.

**Comment by segiddins** [comment]
@homu r+

Thanks, @glennpratt !

**Pull request merged**